### DISCUSSION TITLE
how to replace string before doing a compare or how write my own scripts

### DISCUSSION THREAD

**Discussion by iostrym** [WinMerge/winmerge#2138] [Q&A]
we need to replace wrong end of line char "CRCRLF" by "CRLF" or "CRCR" by " CR" before performing the compare.

I see in in help that there is a script called "Replace" in the list of script "makeupper, makelower, etc" by in my winmerge version I can't see it. the only scripts I have are :
![apply_a_script](https://github.com/WinMerge/winmerge/assets/26224436/aad4bd27-5c90-419e-a044-87040f65b02e)

is there a way also to write my own script to automatically replace "CRCRLF" by "CRLF" without force the user to provide string to be replaced.

**Comment by iostrym** [comment]
hello, here are more detail about my original problem.

here is the file opened by winmerge with corrupted end of line (option activated to display end of line) :
![winmerge_example_corrupted_end_of_line](https://github.com/WinMerge/winmerge/assets/26224436/a422a4c4-94df-42cc-98cc-771048d78bbb)

here is an example of quite the same file without end of line corruption:
![winmerge_example_correct_end_of_line](https://github.com/WinMerge/winmerge/assets/26224436/57d5fee2-d719-4f44-b721-ee26b61b0c18)

using the "replace" function of winmerge I don't know why to type to remove/replace left and down arrow (corrupted end of line)

I can see different symbol for the end of line meaning that there is something strange. using notepad++ for instance I can see that corrupted end of line = CRCRLF (CRLF with LF replaced by CRLF => CRCRLF = root cause of the corruption)
and correct end of line is CRLF (windows)

I know why I have this end of line corruption and there is no way to avoid it. this is why I want to correct it dynamically in winmerge.

thanks in advance for your help

**Comment by iostrym** [comment]
another detail, using an hexa editor I can see that at the end of each line, I have 0xd 0xd 0xa that equal : CR CR LF.

maybe a solution would be to have a plugin in winmerge that would perform hexadecimal replace of "0xd 0xd 0xa" by "0xd 0xa". is it possible to do this ?

**Comment by sdottaka** [comment]
In the latest version ([2.16.36](https://github.com/WinMerge/winmerge/releases/tag/v2.16.36)), you can replace CR+CR with CR using the Replace Unpacker plugin (`Replace -e \r\r \r`) as shown below.



![image](https://github.com/WinMerge/winmerge/assets/98126/a1d7969b-4793-4d0b-9c80-4312256bb66e)
![image](https://github.com/WinMerge/winmerge/assets/98126/95150d9d-2b2e-4d9e-a982-2f0cbd5b4f59)

**Reply by iostrym** [reply]
thanks a lot for your answer. is it possible to launch this replacement after launching the comparison ? or in the command line execution ?
because winmerge is launched by a another tool using a command line.

**Reply by sdottaka** [reply]
Specify the `/unpacker` command line option as below
~~~
"c:\Program Files\WinMerge\WinMergeU.exe" "c:\dev\winmerge-testdata\#2138 how to replace string before doing a compare or how write my own scripts\file1.txt" "c:\dev\winmerge-testdata\#2138 how to replace string before doing a compare or how write my own scripts\file2.txt" /unpacker "Replace -e \r\r \r"
~~~

**Comment by iostrym** [comment]
also I tried the last version of winmerge and I don't have Replace Unpacker plugin. How can I add it ? is it normal I don't have it after installing the last version of the tool ?
![winmerge](https://github.com/WinMerge/winmerge/assets/26224436/10f3cd0f-1efa-4833-87f7-7cb0887e1fa1)

**Reply by sdottaka** [reply]
It seems that only plugins related to the file extensions being compared are displayed. Check the first checkbox as below to display all plugins

![image](https://github.com/WinMerge/winmerge/assets/98126/021460b6-99e3-48e7-9a5d-a175656942c1)

**Comment by iostrym** [comment]
ok thanks no it works, I have the plugin replace displayed. and I was able to perform the replace you proposed.

two problems I have :

1 - replacement is  : \r\r replaced by \r whereas I want to replace only CRCRLF by CRLF. I don't want to replace all successive line return by one line return.
for instance, I tried beyon compare and this tool understand that "CRCRLF" is a strange end of line and display only one end of line with "mixed" type (no problem).
the best would be to be able to perform hexa replacement  : "0xd 0xd 0xa" by "0xd 0xa" unless there is a way to make distinction between a return line in CRLF, CR and LF.

2 - is there a way to launch a compare using command line + the plugin activated with correct replacement ?

**Reply by sdottaka** [reply]
> 1 - replacement is : \r\r replaced by \r whereas I want to replace only CRCRLF by CRLF. I don't want to replace all successive line return by one line return.

Change `Replace -e \r\r \r` to `Replace -e \r\r\n \r\n`. `\r` stands for CR `\n` stands for LF.

> 2 - is there a way to launch a compare using command line + the plugin activated with correct replacement ?

look below
https://github.com/WinMerge/winmerge/discussions/2138#discussioncomment-7868869

**Reply by iostrym** [reply]
thanks.

About \r and \n should it work also in the classic search/replace in winmerge ? if yes, there is a problem because using my corrupted file (added as attachement), I can use \n to search all end of line (CR) and (CRLF) and \r don't display anything.

is it normal ?
[corrupted_line_small.csv](https://github.com/WinMerge/winmerge/files/13705303/corrupted_line_small.csv)

**Reply by sdottaka** [reply]
Unfortunately, it doesn't seem to work if you specify \r.

**Reply by iostrym** [reply]
thanks, it's a bad new. do you think there is another way to do this ? hexadecimal replacement ? others ?

**Reply by sdottaka** [reply]
You can search and replace CR+CR+LF by opening the file in the binary comparison window as shown below. However, there seems to be a problem with the replacement, and the end of the file does not seem to be replaced.

![image](https://github.com/WinMerge/winmerge/assets/98126/4cd3d96f-71a4-44f1-b6d6-afe424611a7e)
![image](https://github.com/WinMerge/winmerge/assets/98126/e016defb-2b25-4220-8fb6-a1fcfa52a5d2)
![image](https://github.com/WinMerge/winmerge/assets/98126/160355da-490d-414d-a100-6d31c72a1a64)

**Comment by iostrym** [comment]
thanks a lot. do you know if it could be asked as an evolution to be able to perform search and replace and doing the correct distinction between CRLF and LF ?

**Reply by sdottaka** [reply]
In commit 7a15b36, it is now possible to search for CR using \r.

**Reply by iostrym** [reply]
Thanks a lot. There is something weird anyway because in brief with your modification:
\r is for CR (mac end of line)
\n is for LF (linux end of line)

It is great for me but there is nothing for "CRLF" that is windows end of line and that is widely used.
Generally tools look for LF and CRLF when we are using \n. 
As I said your correction will help me but it may be strange for other users to have \n looking for LF only and not CRLF.
I know that people could use \r\n to look for CRLF but it is quite unusual to write stuff like this in a regexp.
Maybe there is a need for 3 separated string to look for each:
\r for CR only
\???? For LF only
\n for CR or CRLF 
?

**Reply by sdottaka** [reply]
`\n` does not match CR, but it matches LF of CR+LF, so it is possible to search for CRLF with `\n`

**Reply by iostrym** [reply]
Ok thanks so on my side I could search for \n\n and replace it by \n and my problem is solved, great, I can try it on next winmerge version as it is already corrected, right ?

**Reply by sdottaka** [reply]
Version 2.16.37 Beta has this fix.
https://github.com/WinMerge/winmerge/releases/tag/v2.16.37

**Reply by iostrym** [reply]
correction: if I replace \n\n by \n it won't be ok because I may delete "real blank lines". so I need to replace: 
\r\n by \n
it means to possible things:
1: CRCRLF replaced by CRLF or LF => great !
2: CRLF replaced by CRLF or LF => could happen if there are no CRCRLF string to be replaced. Not wanted but without real consequences.
